tailfs/remote.go

```diff
@@ -0,0 +1,390 @@
+
+package tailfs
+
+import (
+	"bufio"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"net"
+	"net/http"
+	"net/netip"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/tailscale/xnet/webdav"
+	"tailscale.com/safesocket"
+	"tailscale.com/tailfs/compositefs"
+	"tailscale.com/tailfs/webdavfs"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/pathutil"
+)
+
+var (
+	disallowShareAs bool = false
+)
+
+// AllowShareAs indicates whether sharing files as a specific user is allowed
+func AllowShareAs() bool {
+	return !disallowShareAs && doAllowShareAs()
+}
+
+// Share represents a folder that's shared with remote Tailfs nodes.
+type Share struct {
+	// Name is how this share appears on remote nodes.
+	Name string `json:"name"`
+	// Path is the path to the directory on this machine that's being shared.
+	Path string `json:"path"`
+	// As is the UNIX or Windows username of the local account used for this
+	// share. File read/write permissions are enforced based on this username.
+	As string `json:"who"`
+	// Readers is a list of Tailscale principals that are allowed to read this
+	// share.
+	Readers []string `json:"readers,omitempty"`
+	// Writers is a list of Tailscale principals that are allowed to write to
+	// this share.
+	Writers []string `json:"writers,omitempty"`
+}
+
+// ForRemote is the TailFS filesystem exposed to remote nodes. It provides a
+// unified WebDAV interface to local directories that have been shared.
+type ForRemote interface {
+	// SetFileServerAddr sets the address of the file server to which we
+	// should proxy. This is used on platforms like Windows and MacOS
+	// sandboxed where we can't spawn user-specific sub-processes and instead
+	// rely on the UI application that's already running as an unprivileged
+	// user to access the filesystem for us.
+	SetFileServerAddr(addr string)
+
+	// SetShares sets the complete set of shares exposed by this node. If
+	// AllowShareAs() is true, we will use one subprocess per user to access
+	// the filesystem (see userServer). Otherwise, we will use the file server
+	// configured via SetFileServerAddr.
+	SetShares(shares map[string]*Share)
+
+	// ServeHTTP behaves like the similar method from http.Handler but also
+	// accepts a Permissions map that captures the permissions of the connecting
+	// node.
+	ServeHTTP(permissions Permissions, w http.ResponseWriter, r *http.Request)
+
+	// Close() stops serving the WebDAV content
+	Close() error
+}
+
+func NewFileSystemForRemote(logf logger.Logf) ForRemote {
+	fs := &fileSystemForRemote{
+		logf:        logf,
+		lockSystem:  webdav.NewMemLS(),
+		fileSystems: make(map[string]webdav.FileSystem),
+		userServers: make(map[string]*userServer),
+	}
+	return fs
+}
+
+type fileSystemForRemote struct {
+	logf           logger.Logf
+	lockSystem     webdav.LockSystem
+	fileServerAddr string
+	shares         map[string]*Share
+	fileSystems    map[string]webdav.FileSystem
+	userServers    map[string]*userServer
+	mx             sync.RWMutex
```

Comment:
Here and anywhere else relevant, the mutex should precede the fields it guards. Please also document when the caller must hold exclusive and when the caller must hold shared.